src/cargo/core/compiler/timings.rs

```diff
@@ -0,0 +1,641 @@
+            let ctime = self.duration - rmeta_time;
+            let cent = (ctime / self.duration) * 100.0;
+            (rmeta_time, ctime, cent)
+        })
+    }
+
+    fn name_ver(&self) -> String {
+        format!("{} v{}", self.unit.pkg.name(), self.unit.pkg.version())
+    }
+}
+
+// Replace with as_secs_f64 when 1.38 hits stable.
+fn d_as_f64(d: Duration) -> f64 {
+    (d.as_secs() as f64) + f64::from(d.subsec_nanos()) / 1_000_000_000.0
+}
+
+fn render_rustc_info(bcx: &BuildContext<'_, '_>) -> String {
+    let version = bcx
+        .rustc
+        .verbose_version
+        .lines()
+        .next()
+        .expect("rustc version");
+    let requested_target = bcx
+        .build_config
+        .requested_target
+        .as_ref()
+        .map_or("Host", String::as_str);
+    format!(
+        "{}<br>Host: {}<br>Target: {}",
+        version, bcx.rustc.host, requested_target
+    )
+}
+
+static HTML_TMPL: &str = r#"
+<html>
+<head>
+  <title>Cargo Build Timings — {ROOTS}</title>
+  <meta charset="utf-8">
+<style type="text/css">
+html {
+  font-family: sans-serif;
+}
+
+.canvas-container {
+  position: relative;
+  margin-top: 5px;
+  margin-bottom: 5px;
+}
+
+h1 {
+  border-bottom: 1px solid #c0c0c0;
+}
+
+.graph {
+  display: block;
+}
+
+.my-table {
+  margin-top: 20px;
+  margin-bottom: 20px;
+  border-collapse: collapse;
+  box-shadow: 0 5px 10px rgba(0, 0, 0, 0.1);
+}
+
+.my-table th {
+  color: #d5dde5;
+  background: #1b1e24;
+  border-bottom: 4px solid #9ea7af;
+  border-right: 1px solid #343a45;
+  font-size: 18px;
+  font-weight: 100;
+  padding: 12px;
+  text-align: left;
+  vertical-align: middle;
+}
+
+.my-table th:first-child {
+  border-top-left-radius: 3px;
+}
+
+.my-table th:last-child {
+  border-top-right-radius: 3px;
+  border-right:none;
+}
+
+.my-table tr {
+  border-top: 1px solid #c1c3d1;
+  border-bottom: 1px solid #c1c3d1;
+  font-size: 16px;
+  font-weight: normal;
+}
+
+.my-table tr:first-child {
+  border-top:none;
+}
+
+.my-table tr:last-child {
+  border-bottom:none;
+}
+
+.my-table tr:nth-child(odd) td {
+  background: #ebebeb;
+}
+
+.my-table tr:last-child td:first-child {
+  border-bottom-left-radius:3px;
+}
+
+.my-table tr:last-child td:last-child {
+  border-bottom-right-radius:3px;
+}
+
+.my-table td {
+  background: #ffffff;
+  padding: 10px;
+  text-align: left;
+  vertical-align: middle;
+  font-weight: 300;
+  font-size: 14px;
+  border-right: 1px solid #C1C3D1;
+}
+
+.my-table td:last-child {
+  border-right: 0px;
+}
+
+.summary-table td:first-child {
+  vertical-align: top;
+  text-align: right;
+}
+
+.input-table td {
+  text-align: center;
+}
+
+</style>
+</head>
+<body>
+
+<h1>Cargo Build Timings</h1>
+"#;
+
+static HTML_CANVAS: &str = r#"
+<table class="input-table">
+  <tr>
+    <td><label for="min-unit-time">Min unit time:</label></td>
+    <td><label for="scale">Scale:</label></td>
+  </tr>
+  <tr>
+    <td><input type="range" min="0" max="30" value="0" id="min-unit-time"></td>
```

Comment:
FWIW I found this a little unwieldy to work with because it was doing a 1s interval, but could this perhaps be more like a 100ms interval?